modules/wordpress-com-block-editor-iframe.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/**
+ * WordPress.com Block Editor Iframe
+ * Allow new block editor posts to be composed on WordPress.com.
+ * This is auto-loaded as of Jetpack v7.4 for sites connected to WordPress.com only.
+ *
+ * @package Jetpack
+ */
+
+/**
+ * Prevents frame options header from firing if this is a whitelisted iframe request.
+ */
+function jetpack_disable_send_frame_options_header() {
+	if ( jetpack_framing_allowed() ) {
+		remove_action( 'admin_init', 'send_frame_options_header' );
+	}
+}
+add_action( 'admin_init', 'jetpack_disable_send_frame_options_header', 9 );
+
+/**
+ * Adds custom admin body class if this is a whitelisted iframe request.
+ *
+ * @param string $classes Admin body classes.
+ * @return string
+ */
+function jetpack_add_iframed_body_class( $classes ) {
+	if ( jetpack_framing_allowed() ) {
+		$classes .= ' is-iframed ';
+	}
+
+	return $classes;
+}
+add_filter( 'admin_body_class', 'jetpack_add_iframed_body_class' );
+
+/**
+ * Checks whether this is a whitelisted iframe request.
+ *
+ * @return bool
+ */
+function jetpack_framing_allowed() {
+	if ( empty( $_GET['frame-nonce'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification
+		return false;
+	}
+
+	$verified = jetpack_verify_frame_nonce(
+		$_GET['frame-nonce'],  // phpcs:ignore WordPress.Security.NonceVerification
+		'frame-' . Jetpack_Options::get_option( 'id' )
+	);
+
+	if ( $verified && ! defined( 'IFRAME_REQUEST' ) ) {
+		define( 'IFRAME_REQUEST', true );
+	}
+
+	return (bool) $verified;
+}
+
+/**
+ * Verify that correct nonce was used with time limit.
+ *
+ * The user is given an amount of time to use the token, so therefore, since the
+ * UID and $action remain the same, the independent variable is the time.
+ *
+ * @param string $nonce  Nonce that was used in the form to verify.
+ * @param string $action Should give context to what is taking place and be the same when nonce was created.
+ * @return boolean Whether the nonce is valid.
+ */
+function jetpack_verify_frame_nonce( $nonce, $action ) {
+	if ( empty( $nonce ) ) {
+		return false;
+	}
+
+	list( $expiration, $user_id, $hash ) = explode( ':', $nonce, 3 );
+
+	$token = Jetpack_Data::get_access_token( $user_id );
+	if ( ! $token ) {
+		return false;
+	}
+
+	add_filter( 'salt', 'jetpack_filter_salt', 10, 2 );
+	$expected_hash = wp_hash( "$expiration|$action|{$token->external_user_id}", 'jetpack_frame_nonce' );
```

Comment:
We want the signature to be first, and we want to sign using `$user_id`'s token instead of `get_current_user_id()`'s token (otherwise, we can't detect a user ID mismatch - only a signature mismatch).

This code, though uses `get_current_user_id()`'s token.

I'm not sure what the best way is to fix that.

I have a hacky suggestion:

```suggestion
	$expected_hash = wp_hash( "$expiration|$action|{$token->external_user_id}", "jetpack_frame_nonce-$user_id" );
```

And then pull out the `$user_id` from the `$scheme` in `jetpack_filter_salt()` :)

I think any other implementation would have to use a global (not good) or a static variable in a getter/setter function (fine I guess).

Reply:
> This code, though uses `get_current_user_id()`'s token.

Are not we using the `$user_id`'s token here? Or am I missing something?

https://github.com/Automattic/jetpack/pull/11354/files/901ea1e04b0d0119c632daf1f551bfccc7a39857#diff-010062783df39175c5f389986e8d37bfR74

Reply:
It's not being used in the actual signature creation: https://github.com/Automattic/jetpack/pull/11354/files/901ea1e04b0d0119c632daf1f551bfccc7a39857#diff-010062783df39175c5f389986e8d37bfR108

Reply:
Thanks again! I implemented a variation of the getter/setter approach in 9c762e3 which basically moves the logic to a class where we store the `$user_id` passed in the nonce as a property of the class.

Then, the filtered salt uses that property to generate the signature.